{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "92Gw-H6qw1Ou"
      },
      "outputs": [],
      "source": [
        "!pip install anoexpress -q"
      ],
      "id": "92Gw-H6qw1Ou"
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "UGVmVYr89T7i"
      },
      "source": [
        "![image](https://raw.githubusercontent.com/sanjaynagi/AnoExpress/main/docs/logo.png)\n"
      ],
      "id": "UGVmVYr89T7i"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "H68MfuqBOWiH"
      },
      "outputs": [],
      "source": [
        "import anoexpress as xpress\n",
        "import pandas as pd"
      ],
      "id": "H68MfuqBOWiH"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "GdM1MG4woktQ"
      },
      "outputs": [],
      "source": [
        "#@title **AnoExpress** { run: \"auto\" }\n",
        "#@markdown This notebook produces interactive strip and boxplots with plotly, to summarise gene expression across 35 *Anopheles* insecticide resistant v susceptible RNA-Sequencing experiments. There are four different analyses to select\n",
        "#@markdown 'gamb_colu', 'gamb_colu_arab', 'gamb_colu_arab_fun' and 'fun'.\n",
        "#@markdown each which integrates a different combination of Anopheles species – ‘gamb_colu’ (An. gambiae and An. coluzzii), ‘gamb_colu_arab’ (adding An. arabiensis), ‘gamb_colu_arab_fun’ (adding An. funestus), and finally ‘fun’ (An. funestus alone).\n",
        "#@markdown As you include more species, there are less genes overall in the dataset, as we can only include genes with one-to-many orthologs between species.\n",
        "\n",
        "analysis = \"gamb_colu_arab_fun\" #@param ['gamb_colu', 'gamb_colu_arab', 'gamb_colu_arab_fun', 'fun']\n",
        "GeneID = \"AGAP002865\" #@param {type:\"string\"}\n",
        "plot_type = \"strip\" #@param ['strip', 'boxplot']\n",
        "include_microarray_data = True #@param {type:'boolean'}\n",
        "\n",
        "xpress.plot_gene_expression(\n",
        "    analysis=analysis,\n",
        "    gene_id=GeneID,\n",
        "    microarray=include_microarray_data,\n",
        "    title=\"\",\n",
        "    plot_type=plot_type,\n",
        "    height=300)"
      ],
      "id": "GdM1MG4woktQ"
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yqZ2VL2svoDr"
      },
      "source": [
        "We can provide a list of genes, and sort by either AGAP identifier, or by the median fold-change. We can also filter out values that are not significant based on a `pvalue_filter` parameter."
      ],
      "id": "yqZ2VL2svoDr"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Z4oeRiXX_uoi"
      },
      "outputs": [],
      "source": [
        "def summarise_expression(analysis, gene_id, microarray):\n",
        "\n",
        "  # load fc data\n",
        "  fc_data = xpress.data(\n",
        "    analysis=analysis,\n",
        "    data_type='fcs',\n",
        "    gene_id=gene_id,\n",
        "    microarray=microarray,\n",
        "  )\n",
        "  # load pval data\n",
        "  pval_data = xpress.data(\n",
        "    analysis=analysis,\n",
        "    data_type='pvals',\n",
        "    gene_id=gene_id,\n",
        "    microarray=microarray,\n",
        "  )\n",
        "  # load pval data\n",
        "  count_data = xpress.data(\n",
        "    analysis=analysis,\n",
        "    data_type='log2counts',\n",
        "    gene_id=gene_id,\n",
        "    microarray=microarray,\n",
        "  )\n",
        "  # load metadata\n",
        "  metadata = xpress.metadata(analysis=analysis, microarray=microarray)\n",
        "  sample_metadata = xpress.sample_metadata(analysis=analysis)\n",
        "\n",
        "  # add species to pval data and melt\n",
        "  pval_data = pval_data.T.assign(species=metadata.species.to_list())\n",
        "  pval_data = pval_data.reset_index(drop=True)\n",
        "  pval_data = pval_data.melt(id_vars='species', value_name='count')\n",
        "\n",
        "  # add species to fc data and melt\n",
        "  fc_data = fc_data.T.assign(species=metadata.species.to_list())\n",
        "  fc_data = fc_data.reset_index(drop=True)\n",
        "  fc_data = fc_data.melt(id_vars='species', value_name='count')\n",
        "\n",
        "  # add species to fc data and melt\n",
        "  count_data = count_data.T.assign(species=sample_metadata.species.to_list())\n",
        "  count_data = count_data.reset_index(drop=True)\n",
        "  count_data = count_data.melt(id_vars='species', value_name='count')\n",
        "\n",
        "  # get up, down and sig genes\n",
        "  fc_up = fc_data.groupby(['species', 'GeneID'], group_keys=True).apply(lambda x: x > 0)\n",
        "  fc_down = fc_data.groupby(['species', 'GeneID'], group_keys=True).apply(lambda x: x < 0)\n",
        "  pval_sig = pval_data.groupby(['species', 'GeneID'], group_keys=True).apply(lambda x: x < 0.05)\n",
        "\n",
        "  # get intersection of up and sig, down and sig\n",
        "  fc_up_sig = fc_up & pval_sig\n",
        "  fc_down_sig = fc_down & pval_sig\n",
        "\n",
        "  # count total up and down sig\n",
        "  fc_up_sig_summary = fc_up_sig.groupby(['species', 'GeneID']).agg({'count':'sum'}).rename(columns={'count':'up_sig'})\n",
        "  fc_down_sig_summary = fc_down_sig.groupby(['species', 'GeneID']).agg({'count':'sum'}).rename(columns={'count':'down_sig'})\n",
        "  count_summary = fc_up_sig.groupby(['species', 'GeneID']).agg({'count':'count'}).rename(columns={'adj_pval':'total'})\n",
        "\n",
        "  # mean, median, sd per species\n",
        "  median_fc = 2**fc_data.groupby(['species', 'GeneID'], group_keys=True).median().rename(columns={'count':'median_fc'})\n",
        "  mean_fc = 2**fc_data.groupby(['species', 'GeneID'], group_keys=True).mean().rename(columns={'count':'mean_fc'})\n",
        "  std_fc = 2**fc_data.groupby(['species', 'GeneID'], group_keys=True).std().rename(columns={'count':'sd_fc'})\n",
        "\n",
        "  # mean, median, sd counts per species\n",
        "  median_count = 2**count_data.groupby(['species', 'GeneID'], group_keys=True).median().rename(columns={'count':'median_count'})\n",
        "  mean_count = 2**count_data.groupby(['species', 'GeneID'], group_keys=True).mean().rename(columns={'count':'mean_count'})\n",
        "  std_count = 2**count_data.groupby(['species', 'GeneID'], group_keys=True).std().rename(columns={'count':'sd_count'})\n",
        "\n",
        "  # concat\n",
        "  expr_summary = pd.concat([fc_up_sig_summary, fc_down_sig_summary, count_summary, median_fc, mean_fc, std_fc, median_count, mean_count, std_count], axis=1)\n",
        "\n",
        "  return expr_summary.round(2)"
      ],
      "id": "Z4oeRiXX_uoi"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 269
        },
        "id": "8AbkFESxAgJv",
        "outputId": "070c393e-5bfb-4bd2-a456-43dd505d03fb"
      },
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                       up_sig  down_sig  count  median_fc  mean_fc  sd_fc  \\\n",
              "species    GeneID                                                           \n",
              "arabiensis AGAP006227       4         0     15       1.08     1.26   1.56   \n",
              "           AGAP006228       5         0     15       1.15     1.16   1.25   \n",
              "coluzzii   AGAP006227       8         7     31       1.04     1.45   2.30   \n",
              "           AGAP006228      16         6     31       1.32     1.32   1.65   \n",
              "gambiae    AGAP006227       7         0      8       1.65     2.25   1.85   \n",
              "           AGAP006228       3         0      8       1.19     1.15   1.23   \n",
              "\n",
              "                       median_count  mean_count  sd_count  \n",
              "species    GeneID                                          \n",
              "arabiensis AGAP006227        373.51      334.71      1.46  \n",
              "           AGAP006228        760.08      646.63      1.69  \n",
              "coluzzii   AGAP006227        372.22      432.61      2.14  \n",
              "           AGAP006228        754.83      844.19      1.93  \n",
              "gambiae    AGAP006227        319.57      410.59      2.31  \n",
              "           AGAP006228        590.18      563.87      1.26  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-9490cb1a-149a-4014-9a5b-d99a0d920d9c\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th>up_sig</th>\n",
              "      <th>down_sig</th>\n",
              "      <th>count</th>\n",
              "      <th>median_fc</th>\n",
              "      <th>mean_fc</th>\n",
              "      <th>sd_fc</th>\n",
              "      <th>median_count</th>\n",
              "      <th>mean_count</th>\n",
              "      <th>sd_count</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>species</th>\n",
              "      <th>GeneID</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">arabiensis</th>\n",
              "      <th>AGAP006227</th>\n",
              "      <td>4</td>\n",
              "      <td>0</td>\n",
              "      <td>15</td>\n",
              "      <td>1.08</td>\n",
              "      <td>1.26</td>\n",
              "      <td>1.56</td>\n",
              "      <td>373.51</td>\n",
              "      <td>334.71</td>\n",
              "      <td>1.46</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AGAP006228</th>\n",
              "      <td>5</td>\n",
              "      <td>0</td>\n",
              "      <td>15</td>\n",
              "      <td>1.15</td>\n",
              "      <td>1.16</td>\n",
              "      <td>1.25</td>\n",
              "      <td>760.08</td>\n",
              "      <td>646.63</td>\n",
              "      <td>1.69</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">coluzzii</th>\n",
              "      <th>AGAP006227</th>\n",
              "      <td>8</td>\n",
              "      <td>7</td>\n",
              "      <td>31</td>\n",
              "      <td>1.04</td>\n",
              "      <td>1.45</td>\n",
              "      <td>2.30</td>\n",
              "      <td>372.22</td>\n",
              "      <td>432.61</td>\n",
              "      <td>2.14</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AGAP006228</th>\n",
              "      <td>16</td>\n",
              "      <td>6</td>\n",
              "      <td>31</td>\n",
              "      <td>1.32</td>\n",
              "      <td>1.32</td>\n",
              "      <td>1.65</td>\n",
              "      <td>754.83</td>\n",
              "      <td>844.19</td>\n",
              "      <td>1.93</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th rowspan=\"2\" valign=\"top\">gambiae</th>\n",
              "      <th>AGAP006227</th>\n",
              "      <td>7</td>\n",
              "      <td>0</td>\n",
              "      <td>8</td>\n",
              "      <td>1.65</td>\n",
              "      <td>2.25</td>\n",
              "      <td>1.85</td>\n",
              "      <td>319.57</td>\n",
              "      <td>410.59</td>\n",
              "      <td>2.31</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>AGAP006228</th>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>8</td>\n",
              "      <td>1.19</td>\n",
              "      <td>1.15</td>\n",
              "      <td>1.23</td>\n",
              "      <td>590.18</td>\n",
              "      <td>563.87</td>\n",
              "      <td>1.26</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9490cb1a-149a-4014-9a5b-d99a0d920d9c')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-9490cb1a-149a-4014-9a5b-d99a0d920d9c button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-9490cb1a-149a-4014-9a5b-d99a0d920d9c');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-11958f53-0bda-48d4-a07a-7fbc1627c3d0\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-11958f53-0bda-48d4-a07a-7fbc1627c3d0')\"\n",
              "            title=\"Suggest charts.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-11958f53-0bda-48d4-a07a-7fbc1627c3d0 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "    </div>\n",
              "  </div>\n"
            ]
          },
          "metadata": {},
          "execution_count": 41
        }
      ],
      "source": [
        "expr_summary = summarise_expression(analysis='gamb_colu_arab', gene_id=[\"AGAP006227\", \"AGAP006228\"], microarray=True)\n",
        "expr_summary"
      ],
      "id": "8AbkFESxAgJv"
    },
    {
      "cell_type": "code",
      "source": [
        "expr_summary.to_csv(\"coeae1f_2f.expr.summary.tsv\", sep=\"\\t\")"
      ],
      "metadata": {
        "id": "RmnaDs4QcBT7"
      },
      "id": "RmnaDs4QcBT7",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "i84MRbbpGkN4"
      },
      "outputs": [],
      "source": [
        "import plotly.graph_objects as go"
      ],
      "id": "i84MRbbpGkN4"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "background_save": true
        },
        "id": "BzDI3mbwGn3_"
      },
      "outputs": [],
      "source": [
        "def plot_donut_summary(expr_summary, species, GeneID):\n",
        "  values = expr_summary.query(f\"species == '{species}' and GeneID == 'AGAP006227'\")[['up_sig', 'down_sig', 'count']].values[0]\n",
        "  values[2] = values[2]-(values[0] + values[1])\n",
        "  labels = ['over-expressed', 'under-expressed', 'non-significant']\n",
        "  colors = ['gold', 'yellow', 'darkslategray']\n",
        "\n",
        "\n",
        "  # Use `hole` to create a donut-like pie chart\n",
        "  fig = go.Figure(data=[go.Pie(labels=labels, values=values, hole=.4)])\n",
        "  fig.update_traces(hoverinfo='label+percent', textinfo='value', textfont_size=20,\n",
        "                    marker=dict(colors=colors, line=dict(color='#000000', width=2)))\n",
        "  fig.show()"
      ],
      "id": "BzDI3mbwGn3_"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "5ts3kxB4Bpvb"
      },
      "outputs": [],
      "source": [
        "import numpy as np"
      ],
      "id": "5ts3kxB4Bpvb"
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.10"
    },
    "vscode": {
      "interpreter": {
        "hash": "ce681de973941d5edd9bd94c9a2926b7fe65e17e578a68317f38265a230b8ca7"
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}